fix(aio): fine-tune which files are pre-fetched/cached by the ServiceWorker

gkalpak opened the pull request:
_**UPDATED**_
1. Move images that are currently usused to `unused/` sub-directories.
2. Move content-related images to `content/images/` (e.g. `bios`/`guide`/`plunker`).
3. Fix incorrect image URLs in some guides (`ngmodule`, `dependency-injection-in-action`).
4. Update `ngw-manifest.json`:
    - Update the `static.ignore` list.
    - Fix ServiceWorker's routing requests to `index.html`.
    - Cache/pre-fetch certain external resources (that are necessary for displaying the basic pages when offline).
5. Move `/guide/docs.md` to `/marketing/docs.md`, because:
    1. `docs` feels like a "top level" page, similar to `features`, `resources`, `events` etc.
    2. This enables the ServiceWorker to pre-fetch/cache the document (similar to what happens with all other direct children of `content/docs/`), without the need for special-casing it in `ngsw-manifest.json`.
6. Update `@angular/service-worker` to v1.0.0-beta.10 (possibly necessary for external URL caching to work, but a good thing anyway :smiley:)

Fixes #16198.

gkalpak committed:
refactor(aio): move unused images to `unused/` sub-directories

gkalpak committed:
fix(aio): fix ServiceWorker routing

gkalpak committed:
fix(aio): update files ignored by the ServiceWorker

Fixes #16198

gkalpak committed:
feat(aio): cache external resources in the ServiceWorker

Cache resources that are necessary for displaying the basic pages when offline.

gkalpak commented:
@alxhub, I added external URLs in #16318 and I ran into two issues:

1. The requests to external resources have the `?ngsw-cache-bust=...` query param, which in not desirable (e.g. `unpkg.com` returns an `Invalid URL` error). It seems to me that the cache-busting only makes sense for our own resources, not external ones. Do I miss something?

2. The Google Fonts URLs (`https://fonts.googleapis.com/css?family=Droid+Sans+Mono` and `https://fonts.googleapis.com/icon?family=Material+Icons`) fail to be served fromthe cache (although the ServiceWorker does pre-fetch them). I wonder if it is because they have a query param or if it's unrelated.

Any thoughts or feedback are welcome :smiley:

mary-poppins commented:
The angular.io preview for 7b3144b6959bad0c0e739ea0f4a3779b6d80f3b4 is available [here][1].

[1]: https://pr16318-7b3144b6959bad0c0e739ea0f4a3779b6d80f3b4.ngbuilds.io/

IgorMinar commented on aio/src/app/embedded/contributor/contributor.component.ts:
```diff
@@ -57,7 +57,7 @@ import { Contributor } from './contributors.model';
   @Input() person: Contributor;
   noPicture = '_no-one.png';
-  pictureBase = 'content/images/bios/';
+  pictureBase = 'assets/images/bios/';
```
is this right? why don't we keep the bios with the content? we should keep content specific images together with the content that pulls it and use assets only for images used by the shell (not depended upon by any content).

gkalpak replied:
For reference:
We decided to move images that are related to content (not the app shell) to `content/images/`. So, I moved `assets/images/(bios|guide|plunker)` to `content/images/(bios|guide|plunker)`.

IgorMinar commented on aio/ngsw-manifest.json:
```diff
@@ -1,8 +1,8 @@
     "\\.js\\.map$",
-    "^\/assets\/images\/(bios|cookbooks|devguide|examples)\/",
-    "^\/content\/docs\/.*\/"
+    "^/assets/images/(?:bios|guide|plunker|.*/unused)/",
```
see.. we should actually move all the guide and plunker images under `content` and cache all the `assets`, because the stuff left should be all the bits needed by the app shell.

gkalpak committed:
build(aio): update @angular/service-worker to v1.0.0-beta.10

mary-poppins commented:
The angular.io preview for 162482032dc56788f904b0906a890555ffaf7f6c is available [here][1].

[1]: https://pr16318-162482032dc56788f904b0906a890555ffaf7f6c.ngbuilds.io/

gkalpak commented:
WRT https://github.com/angular/angular/pull/16318#issuecomment-297061893:

1. As explained by @alxhub, although there might be cases where the `?ngsw-cache-bust=...` query param is indeed not necessary or desirable for external resources, it is a sane default. Even for external resources, you don't want the browser to serve something that might have been cached before the current `ngsw-manifest.json` was created.
For `unpkg.com` specifically, it works if you explicitly use the URL to the file instead of relying for unpkg to infer it (e.g. use `/lunr@0.7.2/lunr.js` instead of `/lunr@0.7.2`). In any case, we decided to get `lunr` as a local dependency. 

2. With @alxhub's guidance, I got the font URLs caching working by also caching the `.woff2` requests (for the URLs included in the original font files and updating `@angular/service-worker` to v1.0.0-beta.10. (Not 100% sure if updating to beta.10 was necessary.)

gkalpak committed:
refactor(aio): move content-related images to `content/images/`

gkalpak committed:
fix(aio): fix incorrect image URLs in guides

gkalpak committed:
refactor(aio): move `/guide/docs.md` to `/marketing/docs.md`

Because:
1. `docs` feels like a "top level" page, similar to `features`, `resources`, `events` etc.
2. This enables the ServiceWorker to pre-fetch/cache the document (similar to what happens with all
   other direct children of `content/docs/`), without the need for special-casing it in
   `ngsw-manifest.json`.

mary-poppins commented:
The angular.io preview for 646722df0961969e61a742e1933fb9bb493fc98a is available [here][1].

[1]: https://pr16318-646722df0961969e61a742e1933fb9bb493fc98a.ngbuilds.io/

gkalpak commented:
Updated based on feedback (see the updated [description][1]). Probably easier to review per-commit.
@IgorMinar, feel free to TAL :smiley:

Notes:

- I didn't change `lunr` to a local dependency, because I couldn't find a simple/clean way of copying it to dist. The ServiceWorker caches it now, so only the first load (or when no ServiceWorker is available) will be affected. (Still not ideal that an `unpkg.com` outage will break our search for first-time-users. Open to suggestions.)

- There seem to be a few logos inside [assets/images/logos/][2] (which we currently pre-fetch/cache). Do we need them to be pre-fetched/cached?

- @wardbell, I moved `docs.md` from `guide/` back to `marketing/` (mainly because of 5.ii [above][1]), but I can move it back (and have a more complex RegExp in `ngsw-manifest.json`) if you want me to :smiley:

[1]: https://github.com/angular/angular/pull/16318#issue-224170262
[2]: https://github.com/gkalpak/angular/tree/646722df0961969e61a742e1933fb9bb493fc98a/aio/src/assets/images/logos

petebacondarwin commented on aio/ngsw-manifest.json:
```diff
@@ -1,8 +1,8 @@
     "\\.js\\.map$",
-    "^\/assets\/images\/(bios|cookbooks|devguide|examples)\/",
-    "^\/content\/docs\/.*\/"
+    "^/assets/images/.*/unused/",
```
why do we need to ignore these if they are unused?

gkalpak replied:
They are unused, but still present in the `dist/` folder, so SW will try to fetch them. Ideally, I would remove them altogether (and then we wouldn't have to ignore them), but I wasn't sure if someone needed them around.

So, I moved them into `unused` folders and ignored them for now. Someone with more knowledge on whether they are/will be used can verify that they are indeed unnecessary and remove them altogether (but it seemed outside the scope of this PR).

petebacondarwin reviewed (commented):
Some of the images in https://github.com/angular/angular/pull/16318/commits/0895d23a7466bb802bc27513a67bb307641fdfc6 appear to have been modified as well as moved. Do you know why this is?

petebacondarwin commented on aio/content/navigation.json:
```diff
@@ -5,7 +5,7 @@
-      "url": "guide/docs",
+      "url": "docs",
       "title": "Docs",
       "hidden": true
```
this is the `hidden: true` that @wardbell agreed to remove last night.

gkalpak commented:
> Some of the images in 0895d23 appear to have been modified as well as moved. Do you know why this is?

Can you give an example. (I haven't intentionally modified any image.)

petebacondarwin commented:
E.g. https://github.com/angular/angular/pull/16318/files#diff-d33d2bbf86a9a435e51d309d7955313d

gkalpak commented:
For reference (wrt image changes):
GitHub limits the number of renderable files (such as images) to 25. After that, it suppresses the diff (with a "Diff not rendered") message, even if the diff is empty. No images should have changed.

From [the docs][1]:

> The maximum number of renderable files (such as images, PDFs, and GeoJSON files) in a single diff is limited to 25.

[1]: https://help.github.com/articles/limits-for-viewing-content-and-diffs-in-a-repository/#diff-limits
